text/trace/0220-messaging-semantic-conventions-span-structure.md

```diff
@@ -0,0 +1,626 @@
+    direction TB
+    DM1[Deliver m1]-->S1[Settle m1]
+  end
+  PM1-. link .->DM1;
+  PM1-. link .->S1;
+
+  classDef normal fill:green
+  class PM1,DM1,S1 normal
+  linkStyle 0,1,2 color:green,stroke:green
+```
+
+A producer publishes a batch of messages, single messages are delivered to consumers:
+
+```mermaid
+flowchart LR;
+  subgraph PRODUCER
+  direction TB
+  P[Publish]
+  end
+  subgraph CONSUMER
+  direction TB
+  DM1[Deliver m1]
+  DM2[Deliver m2]
+  end
+  P-. link .->DM1;
+  P-. link .->DM2;
+
+  classDef normal fill:green
+  class P,DM1,DM2 normal
+  linkStyle 0,1 color:green,stroke:green
+```
+
+When consuming a single message, the "Deliver" spans can be parented to the creation context:
+
+```mermaid
+flowchart LR;
+  subgraph PRODUCER
+  direction TB
+  P[Publish]
+  end
+  subgraph CONSUMER
+  direction TB
+  DM1[Deliver m1]
+  DM2[Deliver m2]
+  end
+  P-. link .->DM1;
+  P-. link .->DM2;
+  P-->DM1;
+  P-->DM2;
+
+  classDef normal fill:green
+  class P,DM1,DM2 normal
+  linkStyle 0,1,2,3 color:green,stroke:green
+```
+
+A producer creates and publishes a single message, it is delivered as part of a
+batch of message to a consumer:
+
+```mermaid
+flowchart LR;
+  subgraph PRODUCER
+  direction TB
+  PM1[Publish m1]
+  PM2[Publish m2]
+  end
+  subgraph CONSUMER
+  direction TB
+  D[Deliver]-.-PRM1[Process m1]
+  D-.-PRM2[Process m2]
+  end
+  PM1-. link .->D;
+  PM2-. link .->D;
+
+  classDef normal fill:green
+  class PM1,PM2,D normal
+  classDef additional opacity:0.4
+  class PRM1,PRM2 additional
+  linkStyle 0,1 opacity:0.4
+  linkStyle 2,3 color:green,stroke:green
+```
+
+### Pull-based scenarios
+
+A producer creates and publishes a single message, the single message is
+delivered to a consumer. "Create" spans are created independently of the
+"Publish" operation:
+
+```mermaid
+flowchart LR;
+  subgraph PRODUCER
+  direction TB
+  A[Ambient]-->CM1[Create m1]
+  A-->CM2[Create m2]
+  A-->P[Publish]
+  end
+  subgraph CONSUMER
+  direction TB
+  RM1[Receive m1]
+  RM2[Receive m2]
+  end
+  CM1-. link .->RM1;
+  CM2-. link .->RM2;
+  CM1-. link .->P;
+  CM2-. link .->P;
+
+  classDef normal fill:green
+  class CM1,CM2,P,RM1,RM2 normal
+  classDef additional opacity:0.4
+  class A additional
+  linkStyle 0,1,2 opacity:0.4
+  linkStyle 3,4,5,6 color:green,stroke:green
+```
+
+"Create" spans are created as part of the "Publish" operation:
+
+```mermaid
+flowchart LR;
+  subgraph PRODUCER
+  direction TB
+  P[Publish]-->CM1[Create m1]
+  P-->CM2[Create m2]
+  end
+  subgraph CONSUMER
+  direction TB
+  RM1[Receive m1]
+  RM2[Receive m2]
+  end
+  CM1-. link .->RM1;
+  CM2-. link .->RM2;
+
+  classDef normal fill:green
+  class P,CM1,CM2,RM1,RM2 normal
+  linkStyle 0,1,2,3 color:green,stroke:green
+```
+
+A producer creates and publishes a single message, it is delivered as part of a
+batch of message to a consumer. "Process" spans for single messages can be
+created, but are not covered by these conventions:
```

Comment:
Should we also offer a diagram saying that process spans can also be linked to create ones? Also, the "Push" based doesn't have such diagrams, I wonder if should have the same diagrams/cases in both to be consistent? I say because one might think that "process" can only be instrumented in pull based, since there's no example of it in the push-based section.

Reply:
> Should we also offer a diagram saying that process spans can also be linked to create ones?

No, I'd like to avoid showing those links when they are not part of conventions. I'd like to avoid the situation we have with the current semantic conventions, where people are just inferring things from examples that are otherwise unspecified.

> Also, the "Push" based doesn't have such diagrams, I wonder if should have the same diagrams/cases in both to be consistent? 

I think this diagram covers a similar case for the "Push" based scenarios:
![image](https://user-images.githubusercontent.com/16337442/231586574-de82ecde-f0ad-42b5-8c98-90524f581fbf.png)

Reply:
@joaopgrassi  Please resolve this conversation if the diagram shown above covers the scenario you had in mind.

Reply:
All good!